File: dart_sample/lib/demos/material/refresh_indicator_demo_page.dart
import 'package:flutter/material.dart';

class RefreshIndicatorDemoPage extends StatefulWidget {
  const RefreshIndicatorDemoPage({super.key});

  @override
  State<RefreshIndicatorDemoPage> createState() =>
      _RefreshIndicatorDemoPageState();
}

class _RefreshIndicatorDemoPageState extends State<RefreshIndicatorDemoPage> {
  int _variant = 0;
  bool _useCupertinoPlatform = false;
  bool _useThemeOverrides = false;
  bool _useSchemeColor = false;
  int _refreshCount = 0;
  String _status = 'idle';

  @override
  Widget build(BuildContext context) {
    final ThemeData baseTheme = Theme.of(context);
    final ThemeData theme = baseTheme.copyWith(
      platform: _useCupertinoPlatform
          ? TargetPlatform.iOS
          : TargetPlatform.android,
      primaryColor: const Color(0xFFFF6F00),
      colorScheme: baseTheme.colorScheme.copyWith(
        primary: const Color(0xFF00897B),
      ),
      progressIndicatorTheme: _useThemeOverrides
          ? const ProgressIndicatorThemeData(
              color: Color(0xFF6A1B9A),
              refreshBackgroundColor: Color(0xFFFFF3E0),
              strokeAlign: -1,
              strokeCap: StrokeCap.round,
            )
          : const ProgressIndicatorThemeData(),
    );

    return Theme(
      data: theme,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          const Text(
            'RefreshIndicator + RefreshProgressIndicator',
            style: TextStyle(fontSize: 20, color: Colors.black),
          ),
          const SizedBox(height: 8),
          const Text(
            'Pull the list down from its top edge. Cycle Material/adaptive/no-spinner paths and theme the refresh surface.',
            style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
          ),
          const SizedBox(height: 10),
          Row(
            children: <Widget>[
              _buildButton(_variantLabel(), () {
                setState(() => _variant = (_variant + 1) % 3);
              }, 126),
              const SizedBox(width: 8),
              _buildButton(
                _useCupertinoPlatform ? 'platform=iOS' : 'platform=Android',
                () => setState(
                  () => _useCupertinoPlatform = !_useCupertinoPlatform,
                ),
                132,
              ),
              const SizedBox(width: 8),
              _buildButton(
                _useThemeOverrides ? 'theme=on' : 'theme=off',
                () => setState(() => _useThemeOverrides = !_useThemeOverrides),
                104,
              ),
            ],
          ),
          const SizedBox(height: 8),
          Row(
            children: <Widget>[
              _buildButton(
                _useSchemeColor ? 'color=scheme' : 'color=widget',
                () => setState(() => _useSchemeColor = !_useSchemeColor),
                126,
              ),
              const SizedBox(width: 8),
              const Text(
                'scheme teal; legacy primary orange',
                style: TextStyle(fontSize: 12, color: Color(0xFF607D8B)),
              ),
            ],
          ),
          const SizedBox(height: 8),
          Text(
            'status=$_status, refreshCount=$_refreshCount; '
            'drag past the armed threshold, then release',
            style: const TextStyle(fontSize: 12, color: Color(0xFF607D8B)),
          ),
          const SizedBox(height: 10),
          Expanded(child: _buildRefreshWrapper(_buildList())),
        ],
      ),
    );
  }

  Widget _buildRefreshWrapper(Widget child) {
    switch (_variant) {
      case 1:
        return RefreshIndicator.adaptive(
          onRefresh: _handleRefresh,
          color: _useSchemeColor ? null : const Color(0xFF1565C0),
          semanticsLabel: 'Refresh sample list',
          child: child,
        );
      case 2:
        return RefreshIndicator.noSpinner(
          onRefresh: _handleRefresh,
          onStatusChange: (RefreshIndicatorStatus? status) {
            setState(() => _status = status?.name ?? 'idle');
          },
          semanticsLabel: 'Refresh sample list',
          child: child,
        );
      default:
        return RefreshIndicator(
          onRefresh: _handleRefresh,
          color: _useSchemeColor ? null : const Color(0xFF1565C0),
          backgroundColor: _useThemeOverrides ? null : Colors.white,
          semanticsLabel: 'Refresh sample list',
          child: child,
        );
    }
  }

  Widget _buildList() {
    return ListView.builder(
      itemCount: 24,
      itemExtent: 54,
      padding: const EdgeInsets.all(8),
      itemBuilder: (BuildContext context, int index) => Container(
        color: index.isEven ? Colors.white : const Color(0xFFF5F7FA),
        padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 10),
        child: Text(
          'refresh row #${index + 1}',
          style: const TextStyle(fontSize: 13, color: Colors.black),
        ),
      ),
    );
  }

  Future<void> _handleRefresh() async {
    if (mounted) setState(() => _status = 'refresh');
    await Future<void>.delayed(const Duration(milliseconds: 650));
    if (mounted) {
      setState(() {
        _refreshCount += 1;
        _status = 'done';
      });
    }
  }

  String _variantLabel() => switch (_variant) {
    1 => 'adaptive',
    2 => 'noSpinner',
    _ => 'material',
  };

  Widget _buildButton(String label, VoidCallback onTap, double width) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onTap,
        style: TextButton.styleFrom(
          minimumSize: const Size(0, 36),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          backgroundColor: const Color(0xFFE9F0FF),
          foregroundColor: Colors.black,
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/Material/RefreshIndicatorDemoPage.cs
using System;
using System.Threading.Tasks;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/material/refresh_indicator_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class RefreshIndicatorDemoPage : StatefulWidget
{
    public override State CreateState() => new RefreshIndicatorDemoPageState();
}

internal sealed class RefreshIndicatorDemoPageState : State
{
    private int _variant;
    private bool _useCupertinoPlatform;
    private bool _useThemeOverrides;
    private bool _useSchemeColor;
    private int _refreshCount;
    private string _status = "idle";

    public override Widget Build(BuildContext context)
    {
        var baseTheme = Theme.Of(context);
        var theme = baseTheme with
        {
            Platform = _useCupertinoPlatform ? TargetPlatform.IOS : TargetPlatform.Android,
            PrimaryColor = Color.Parse("#FFFF6F00"),
            ColorScheme = baseTheme.ColorScheme.CopyWith(primary: Color.Parse("#FF00897B")),
            ProgressIndicatorTheme = _useThemeOverrides
                ? new ProgressIndicatorThemeData(
                    Color: Color.Parse("#FF6A1B9A"),
                    RefreshBackgroundColor: Color.Parse("#FFFFF3E0"),
                    StrokeAlign: -1,
                    StrokeCap: StrokeCap.Round)
                : new ProgressIndicatorThemeData()
        };

        return new Theme(
            data: theme,
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 10,
                children:
                [
                    new Text("RefreshIndicator + RefreshProgressIndicator", fontSize: 20, color: Colors.Black),
                    new Text(
                        "Pull the list down from its top edge. Cycle Material/adaptive/no-spinner paths and theme the refresh surface.",
                        fontSize: 14,
                        color: Color.Parse("#8A000000")),
                    new Row(
                        spacing: 8,
                        children:
                        [
                            BuildButton(VariantLabel(), () => SetState(() => _variant = (_variant + 1) % 3), 126),
                            BuildButton(_useCupertinoPlatform ? "platform=iOS" : "platform=Android", () => SetState(() => _useCupertinoPlatform = !_useCupertinoPlatform), 132),
                            BuildButton(_useThemeOverrides ? "theme=on" : "theme=off", () => SetState(() => _useThemeOverrides = !_useThemeOverrides), 104),
                        ]),
                    new Row(
                        children:
                        [
                            BuildButton(
                                _useSchemeColor ? "color=scheme" : "color=widget",
                                () => SetState(() => _useSchemeColor = !_useSchemeColor),
                                126),
                            new Text(
                                "scheme teal; legacy primary orange",
                                fontSize: 12,
                                color: Color.Parse("#FF607D8B")),
                        ]),
                    new Text(
                        $"status={_status}, refreshCount={_refreshCount}; drag past the armed threshold, then release",
                        fontSize: 12,
                        color: Color.Parse("#FF607D8B")),
                    new Expanded(child: BuildRefreshWrapper(BuildList())),
                ]));
    }

    private Widget BuildRefreshWrapper(Widget child)
    {
        return _variant switch
        {
            1 => RefreshIndicator.Adaptive(
                onRefresh: HandleRefresh,
                child: child,
                color: _useSchemeColor ? null : Color.Parse("#FF1565C0"),
                semanticsLabel: "Refresh sample list"),
            2 => RefreshIndicator.NoSpinner(
                onRefresh: HandleRefresh,
                onStatusChange: status => SetState(() => _status = status?.ToString().ToLowerInvariant() ?? "idle"),
                child: child,
                semanticsLabel: "Refresh sample list"),
            _ => new RefreshIndicator(
                onRefresh: HandleRefresh,
                child: child,
                color: _useSchemeColor ? null : Color.Parse("#FF1565C0"),
                backgroundColor: _useThemeOverrides ? null : Colors.White,
                semanticsLabel: "Refresh sample list"),
        };
    }

    private Widget BuildList()
    {
        return ListView.Builder(
            itemCount: 24,
            itemExtent: 54,
            padding: new Thickness(8),
            addAutomaticKeepAlives: false,
            itemBuilder: (_, index) => new Container(
                color: index % 2 == 0 ? Colors.White : Color.Parse("#FFF5F7FA"),
                padding: new Thickness(12, 10),
                child: new Text($"refresh row #{index + 1}", fontSize: 13, color: Colors.Black)));
    }

    private async Task HandleRefresh()
    {
        if (Mounted) SetState(() => _status = "refresh");
        await Task.Delay(650);
        if (Mounted)
        {
            SetState(() =>
            {
                _refreshCount += 1;
                _status = "done";
            });
        }
    }

    private string VariantLabel() => _variant switch
    {
        1 => "adaptive",
        2 => "noSpinner",
        _ => "material",
    };

    private static Widget BuildButton(string label, Action onTap, double width) =>
        new SizedBox(
            width: width,
            child: new TextButton(
                onPressed: onTap,
                child: new Text(label, fontSize: 12),
                style: TextButton.StyleFrom(
                    foregroundColor: Colors.Black,
                    backgroundColor: Color.Parse("#FFE9F0FF"),
                    padding: new Thickness(10, 8),
                    minimumSize: new Size(64, 36),
                    shape: new RoundedRectangleBorder(
                        borderRadius: BorderRadius.Circular(8)))));
}
